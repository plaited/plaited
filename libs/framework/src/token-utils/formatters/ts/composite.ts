import { camelCase } from '@plaited/utils'
import { Formatter, CompositeToken } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar, isValidAlias } from '../ts-utils.js'

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value, allTokens)}`
  }
  const toRet: string[] = []
  for (const key in $value) {
    const val = $value[key]
    if (isValidAlias(val, allTokens)) {
      toRet.push(`  ${key}: ${resolveTSVar(val, allTokens)},`)
    }
  }
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
