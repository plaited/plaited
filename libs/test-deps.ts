// -- std --
export {
  assert,
  assertEquals,
  assertFalse,
} from 'https://deno.land/std@0.176.0/testing/asserts.ts'
export { assertSnapshot } from 'https://deno.land/std@0.176.0/testing/snapshot.ts'
export {
  afterEach,
  beforeEach,
  describe,
  it,
} from 'https://deno.land/std@0.176.0/testing/bdd.ts'
export { resolve } from 'https://deno.land/std@0.175.0/path/mod.ts'
// -- sinon --
export { default as sinon } from 'https://esm.sh/sinon@15.0.1'

// -- beautify --
export { default as beautify } from 'https://esm.sh/beautify@0.0.8'
