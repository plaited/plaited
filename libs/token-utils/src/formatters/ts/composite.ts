import { camelCase } from '@plaited/utils'
import { Formatter, AliasValue, FlexToken, GridToken, TypographyToken } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar } from '../ts-utils.js'

type CompositeToken = FlexToken | GridToken | TypographyToken

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value as AliasValue, allTokens)}`
  }
  const toRet = Object.entries($value).map(([key, val]) => `  ${key}: ${resolveTSVar(val, allTokens)},`)
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
