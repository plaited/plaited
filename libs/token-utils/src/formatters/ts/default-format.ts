import { camelCase } from '@plaited/utils'
import { Formatter } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { isValidAlias, getCssVar } from '../ts-utils.js'

export const defaultFormat: Formatter = (token, { allTokens, tokenPath }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  return `export const ${camelCase(tokenPath.join(' '))} = ${getCssVar(tokenPath)}`
}
