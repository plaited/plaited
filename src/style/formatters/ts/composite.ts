import type { CompositeToken } from '../../token.types.js'
import type { Formatter } from '../formatters.types.js'
import { camelCase } from '../../../utils/case.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar, isValidAlias } from '../ts-utils.js'

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value, allTokens)}`
  }
  const toRet: string[] = []
  for (const key in $value) {
    const val = $value[key]
    if (isValidAlias(val, allTokens)) {
      toRet.push(`  ${key}: ${resolveTSVar(val, allTokens)},`)
    }
  }
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
