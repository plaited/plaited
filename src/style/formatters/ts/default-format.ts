import type { Formatter } from '../formatters.types.ts'
import { camelCase } from '../../../utils/case.ts'
import { hasAlias } from '../has-alias.ts'
import { isValidAlias, getCssVar } from '../ts-utils.ts'

export const defaultFormat: Formatter = (token, { allTokens, tokenPath }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  return `export const ${camelCase(tokenPath.join(' '))} = ${getCssVar(tokenPath)}`
}
