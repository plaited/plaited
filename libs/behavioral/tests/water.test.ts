import { assertEquals, assertSnapshot } from '../../test-deps.ts'
import { bProgram, DevCallback } from '../mod.ts'

Deno.test('Add hot water 3 times', () => {
  const actual: string[] = []
  const { addThreads, thread, sync, trigger, feedback } = bProgram()
  addThreads({
    addHot: thread(
      sync({ request: { event: 'hot' } }),
      sync({ request: { event: 'hot' } }),
      sync({ request: { event: 'hot' } }),
    ),
  })
  feedback({
    hot() {
      actual.push('hot')
    },
  })
  trigger({ event: 'start' })
  assertEquals(actual, ['hot', 'hot', 'hot'])
})

Deno.test('Add hot/cold water 3 times', () => {
  const actual: string[] = []
  const { addThreads, thread, sync, trigger, feedback } = bProgram()
  addThreads({
    addHot: thread(
      sync({ request: { event: 'hot' } }),
      sync({ request: { event: 'hot' } }),
      sync({ request: { event: 'hot' } }),
    ),
    addCold: thread(
      sync({ request: { event: 'cold' } }),
      sync({ request: { event: 'cold' } }),
      sync({ request: { event: 'cold' } }),
    ),
  })
  feedback({
    hot() {
      actual.push('hot')
    },
    cold() {
      actual.push('cold')
    },
  })
  trigger({ event: 'start' })
  assertEquals(actual, [
    'hot',
    'hot',
    'hot',
    'cold',
    'cold',
    'cold',
  ])
})

Deno.test('interleave', () => {
  const actual: string[] = []
  const { addThreads, thread, sync, trigger, feedback, loop } = bProgram()
  addThreads({
    addHot: thread(
      sync({ request: { event: 'hot' } }),
      sync({ request: { event: 'hot' } }),
      sync({ request: { event: 'hot' } }),
    ),
    addCold: thread(
      sync({ request: { event: 'cold' } }),
      sync({ request: { event: 'cold' } }),
      sync({ request: { event: 'cold' } }),
    ),
    mixHotCold: loop([
      sync({
        waitFor: { event: 'hot' },
        block: { event: 'cold' },
      }),
      sync({
        waitFor: { event: 'cold' },
        block: { event: 'hot' },
      }),
    ]),
  })
  feedback({
    hot() {
      actual.push('hot')
    },
    cold() {
      actual.push('cold')
    },
  })
  trigger({ event: 'start' })
  assertEquals(actual, [
    'hot',
    'cold',
    'hot',
    'cold',
    'hot',
    'cold',
  ])
})

Deno.test('logging', (t) => {
  const logs: Parameters<DevCallback>[0][] = []
  const { addThreads, thread, sync, trigger, loop } = bProgram({
    dev: (msg) => logs.push(msg),
  })
  addThreads({
    addHot: thread(
      sync({ request: { event: 'hot' } }),
      sync({ request: { event: 'hot' } }),
      sync({ request: { event: 'hot' } }),
    ),
    addCold: thread(
      sync({ request: { event: 'cold' } }),
      sync({ request: { event: 'cold' } }),
      sync({ request: { event: 'cold' } }),
    ),
    mixHotCold: loop([
      sync({
        waitFor: { event: 'hot' },
        block: { event: 'cold' },
      }),
      sync({
        waitFor: { event: 'cold' },
        block: { event: 'hot' },
      }),
    ]),
  })
  trigger({ event: 'start' })
  assertSnapshot(t, logs)
})
