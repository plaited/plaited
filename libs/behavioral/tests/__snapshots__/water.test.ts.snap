export const snapshot = {};

snapshot[`logging 1`] = `
[
  [
    {
      priority: 0,
      request: [
        {
          detail: undefined,
          event: "start",
        },
      ],
      selected: true,
      thread: "start",
      trigger: true,
      waitFor: [
        {
          cb: [Function: cb],
          event: "",
        },
      ],
    },
    {
      priority: 1,
      request: [
        {
          event: "hot",
        },
      ],
      thread: "addHot",
    },
    {
      priority: 2,
      request: [
        {
          event: "cold",
        },
      ],
      thread: "addCold",
    },
    {
      block: [
        {
          event: "cold",
        },
      ],
      priority: 3,
      thread: "mixHotCold",
      waitFor: [
        {
          event: "hot",
        },
      ],
    },
  ],
  [
    {
      priority: 1,
      request: [
        {
          event: "hot",
        },
      ],
      selected: true,
      thread: "addHot",
    },
    {
      priority: 2,
      request: [
        {
          event: "cold",
        },
      ],
      thread: "addCold",
    },
    {
      block: [
        {
          event: "cold",
        },
      ],
      priority: 3,
      thread: "mixHotCold",
      waitFor: [
        {
          event: "hot",
        },
      ],
    },
  ],
  [
    {
      priority: 1,
      request: [
        {
          event: "hot",
        },
      ],
      thread: "addHot",
    },
    {
      priority: 2,
      request: [
        {
          event: "cold",
        },
      ],
      selected: true,
      thread: "addCold",
    },
    {
      block: [
        {
          event: "hot",
        },
      ],
      priority: 3,
      thread: "mixHotCold",
      waitFor: [
        {
          event: "cold",
        },
      ],
    },
  ],
  [
    {
      priority: 1,
      request: [
        {
          event: "hot",
        },
      ],
      selected: true,
      thread: "addHot",
    },
    {
      priority: 2,
      request: [
        {
          event: "cold",
        },
      ],
      thread: "addCold",
    },
    {
      block: [
        {
          event: "cold",
        },
      ],
      priority: 3,
      thread: "mixHotCold",
      waitFor: [
        {
          event: "hot",
        },
      ],
    },
  ],
  [
    {
      priority: 1,
      request: [
        {
          event: "hot",
        },
      ],
      thread: "addHot",
    },
    {
      priority: 2,
      request: [
        {
          event: "cold",
        },
      ],
      selected: true,
      thread: "addCold",
    },
    {
      block: [
        {
          event: "hot",
        },
      ],
      priority: 3,
      thread: "mixHotCold",
      waitFor: [
        {
          event: "cold",
        },
      ],
    },
  ],
  [
    {
      priority: 1,
      request: [
        {
          event: "hot",
        },
      ],
      selected: true,
      thread: "addHot",
    },
    {
      priority: 2,
      request: [
        {
          event: "cold",
        },
      ],
      thread: "addCold",
    },
    {
      block: [
        {
          event: "cold",
        },
      ],
      priority: 3,
      thread: "mixHotCold",
      waitFor: [
        {
          event: "hot",
        },
      ],
    },
  ],
  [
    {
      priority: 2,
      request: [
        {
          event: "cold",
        },
      ],
      selected: true,
      thread: "addCold",
    },
    {
      block: [
        {
          event: "hot",
        },
      ],
      priority: 3,
      thread: "mixHotCold",
      waitFor: [
        {
          event: "cold",
        },
      ],
    },
  ],
]
`;
